File: src/main/java/AoC2022_24.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import java.math.BigInteger;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.navigation.Heading;
import com.github.pareronia.aoc.navigation.Headings;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

import lombok.EqualsAndHashCode;
import lombok.RequiredArgsConstructor;

public class AoC2022_24 extends AoCBase {
    private static final Set<Heading> DIRS = Set.of(
        Headings.NORTH.get(), Headings.EAST.get(), Headings.SOUTH.get(),
        Headings.WEST.get(), Heading.of(0, 0)
    );
    private static final Map<Character, Heading> BLIZZARD_DIRS = Map.of(
        '^', Headings.NORTH.get(), '>', Headings.EAST.get(),
        'v', Headings.SOUTH.get(), '<', Headings.WEST.get()
    );
    
    private final boolean[][][] blizzardsByTime;
    private final Position start;
    private final Position end;
    private final int width;
    private final int height;
    private final int period;
    
    private AoC2022_24(final List<String> input, final boolean debug) {
        super(debug);
        this.width = input.get(0).length() - 2;
        this.height = input.size() - 2;
        this.period = height * (width /
            BigInteger.valueOf(height).gcd(BigInteger.valueOf(width)).intValue());
        this.start = Position.of(0, height);
        this.end = Position.of(width - 1, -1);
        final List<Blizzard> blizzards = new ArrayList<>();
        IntStream.range(0, input.size()).forEach(y -> {
            for (int x = 0; x < input.get(y).length(); x++) {
                final char ch = input.get(y).charAt(x);
                if (ch != '.' && ch != '#') {
                    final Position position = Position.of(x - 1,  height - y);
                    blizzards.add(new Blizzard(position, BLIZZARD_DIRS.get(ch)));
                }
            }
        });
        this.blizzardsByTime = new boolean[this.period][this.width][this.height];
        for (final int i : range(this.period)) {
            for (final Blizzard blizzard : blizzards) {
                final Position b = blizzard.at(i, this.width, this.height);
                this.blizzardsByTime[i][b.getX()][b.getY()] = true;
            }
        }
        log("width: " + this.width);
        log("height: " + this.height);
        log("period: " + this.period);
        log("start: " + this.start);
        log("end: " + this.end);
    }
    
    public static final AoC2022_24 create(final List<String> input) {
        return new AoC2022_24(input, false);
    }

    public static final AoC2022_24 createDebug(final List<String> input) {
        return new AoC2022_24(input, true);
    }
    
    private boolean isBlizzard(final Position position, final int time) {
        return this.blizzardsByTime[time % this.period][position.getX()][position.getY()];
    }

    private boolean inBounds(final Position p) {
        return 0 <= p.getX() && p.getX() < this.width
            && 0 <= p.getY() && p.getY() < this.height;
    }
    
    @RequiredArgsConstructor
    @EqualsAndHashCode
    private static final class State {
        private final int time;
        private final Position position;
    }
    
    private int solve(final Position start, final Position end, final int startTime) {
        final Deque<State> q = new ArrayDeque<>();
        q.add(new State(startTime, start));
        final Set<State> seen = new HashSet<>();
        int time = -1;
        while (!q.isEmpty()) {
            final State state = q.poll();
            if (state.position.equals(end)) {
                log("time: " + state.time);
                return state.time;
            }
            if (time < state.time) {
                time = state.time;
                seen.clear();
            }
            final int nextTime = state.time + 1;
            for (final Heading d : DIRS) {
                final Position n = state.position.translate(d);
                if (n.equals(start) || n.equals(end)
                        || inBounds(n) && !isBlizzard(n, nextTime)
                ) {
                    final State newState = new State(nextTime, n);
                    if (!seen.contains(newState)) {
                        seen.add(newState);
                        q.add(newState);
                    }
                }
            }
        }
        throw new IllegalStateException("Unsolvable");
    }

    @Override
    public Integer solvePart1() {
        return solve(this.start, this.end, 0);
    }

    @Override
    public Integer solvePart2() {
        final int there = solve(this.start, this.end, 0);
        final int thereAndBack = solve(this.end, this.start, there);
        final int thereAndBackAndThereAgain = solve(this.start, this.end, thereAndBack);
        return thereAndBackAndThereAgain;
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_24.createDebug(TEST).solvePart1() == 18;
        assert AoC2022_24.createDebug(TEST).solvePart2() == 54;

        final Puzzle puzzle = Aocd.puzzle(2022, 24);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_24.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_24.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST = splitLines(
        "#.######\r\n" +
        "#>>.<^<#\r\n" +
        "#.<..<<#\r\n" +
        "#>v.><>#\r\n" +
        "#<^v^^>#\r\n" +
        "######.#"
    );
    
    @RequiredArgsConstructor
    private static final class Blizzard {
        private final Position position;
        private final Heading heading;
        
        public Position at(final int tick, final int width, final int height) {
            final int amount = horizontal() ? tick % width : tick % height;
            final Position newPosition
                    = this.position.translate(this.heading, amount);
            final Position pp = Position.of(
                    (width + newPosition.getX()) % width,
                    (height + newPosition.getY()) % height);
            assert 0 <= pp.getX() && pp.getX() < width
                    && 0 <= pp.getY() && pp.getY() < height;
            return pp;
        }
        
        private boolean horizontal() {
            return this.heading.equals(Headings.EAST.get())
                    || this.heading.equals(Headings.WEST.get());
        }
    }
}

File: src/main/cpp/2022/24/AoC2022_24.cpp
#include <assert.h>

#include <algorithm>
#include <deque>
#include <numeric>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;
using Position = pair<int, int>;
using Direction = pair<int, int>;
using State = tuple<int, int, int>;
using Blizzards = unordered_map<int, unordered_set<Position>>;

#define Position(x, y) make_pair(x, y)
#define Direction(x, y) make_pair(x, y)
#define State(position, time) make_tuple(position.first, position.second, time)

const Direction N = Direction(0, 1);
const Direction S = Direction(0, -1);
const Direction W = Direction(-1, 0);
const Direction E = Direction(1, 0);
const unordered_set<Direction> ALL_DIRS = {N, E, S, W, Direction(0, 0)};
const unordered_map<char, Direction> BLIZZARD_DIRS = {
    {'^', N}, {'>', E}, {'v', S}, {'<', W}};

struct Blizzard {
    const Position position;
    const Direction direction;

    Blizzard(const Position& position, const Direction& direction)
        : position(position), direction(direction) {}
    Position at(const int time, const int width, const int height) const {
        int amount;
        if (this->direction == W || this->direction == E) {
            amount = time % width;
        } else if (this->direction == N || this->direction == S) {
            amount = time % height;
        } else {
            assert(false);
        }
        const Position newPosition(
            this->position.first + amount * this->direction.first,
            this->position.second + amount * this->direction.second);
        const Position pp((width + newPosition.first) % width,
                          (height + newPosition.second) % height);
        assert(0 <= pp.first && pp.first < width && 0 <= pp.second &&
               pp.second < height);
        return pp;
    }
};

struct Valley {
    const int height;
    const int width;
    const int period;
    const Position start;
    const Position end;
    const Blizzards blizzards;

    Valley(const int height, const int width, const int period,
           const Blizzards blizzards)
        : height(height),
          width(width),
          period(period),
          start(Position(0, height)),
          end(Position(width - 1, -1)),
          blizzards(blizzards) {}
    bool isBlizzard(const Position& position, const int time) const {
        const unordered_set<Position>& blizzardsAtTime =
            this->blizzards.at(time % this->period);
        return blizzardsAtTime.find(position) != blizzardsAtTime.end();
    }
    bool inBounds(const Position& position) const {
        return 0 <= position.first && position.first < this->width &&
               0 <= position.second && position.second < this->height;
    }
};

Valley parse(const vector<string>& input) {
    const int height = input.size() - 2;
    const int width = input[0].size() - 2;
    const int period = lcm(height, width);
    vector<Blizzard> blizzards;
    for (uint y = 0; y < input.size(); y++) {
        for (uint x = 0; x < input[y].size(); x++) {
            const char ch = input[y][x];
            if (ch != '.' && ch != '#') {
                const Position position(x - 1, height - y);
                blizzards.push_back(Blizzard(position, BLIZZARD_DIRS.at(ch)));
            }
        }
    }
    Blizzards blizzardsByTime;
    for (int i = 0; i < period; i++) {
        blizzardsByTime.try_emplace(i, unordered_set<Position>());
        for (const Blizzard& blizzard : blizzards) {
            blizzardsByTime.at(i).insert(blizzard.at(i, width, height));
        }
    }
    Valley valley(height, width, period, blizzardsByTime);
    return valley;
}

template <>
struct std::hash<State> {
    size_t operator()(const State& p) const {
        return get<0>(p) + 31 * (get<1>(p) << 1) + 73 * (get<2>(p) << 3);
    }
};

int solve(const Valley& valley, const Position& start, const Position& end,
          const int startTime) {
    deque<State> q;
    q.push_back(State(start, startTime));
    unordered_set<State> seen;
    while (q.size() > 0) {
        const State& state = q.front();
        q.pop_front();
        const int x = get<0>(state);
        const int y = get<1>(state);
        const int time = get<2>(state);
        if (x == end.first && y == end.second) {
            return time;
        }
        const int nextTime = time + 1;
        for (const Direction& d : ALL_DIRS) {
            const Position n(x + d.first, y + d.second);
            if (n == start || n == end ||
                (valley.inBounds(n) && !valley.isBlizzard(n, nextTime))) {
                const State newState = State(n, nextTime);
                if (seen.find(newState) == seen.end()) {
                    seen.insert(newState);
                    q.push_back(newState);
                }
            }
        }
    }
    throw "Unsolvable";
}

int part1(const vector<string>& input) {
    const Valley& valley = parse(input);
    return solve(valley, valley.start, valley.end, 0);
}

int part2(const vector<string>& input) {
    const Valley& valley = parse(input);
    const int there = solve(valley, valley.start, valley.end, 0);
    const int thereAndBack = solve(valley, valley.end, valley.start, there);
    const int thereAndBackAndThereAgain =
        solve(valley, valley.start, valley.end, thereAndBack);
    return thereAndBackAndThereAgain;
}

// clang-format off
const vector<string> TEST = {
"#.######",
"#>>.<^<#",
"#.<..<<#",
"#>v.><>#",
"#<^v^^>#",
"######.#"
};
// clang-format on

void samples() {
    assert(part1(TEST) == 18);
    assert(part2(TEST) == 54);
}

MAIN(2022, 24)
